jobclient/python/groups.py

```diff
@@ -0,0 +1,175 @@
+        :param parameters: Parameters, defaults to an empty `dict`.
+        :type parameters: Dict[str, str], optional
+        """
+        self.placement_type = HostPlacementType
+        self.parameters = parameters
+
+    def parameter(self, key: str) -> str:
+        """Get the value associated with a parameter."""
+        return self.parameters[key]
+
+    def to_dict(self) -> dict:
+        """Generate this object's `dict` representation."""
+        return {
+            'type': self.placement_type,
+            'parameters': self.parameters
+        }
+
+    def __index__(self, key: str) -> str:
+        """Get the value associated with a parameter."""
+        return self.parameter(key)
+
+
+class Group:
+    """A group of linked jobs."""
+    uuid: UUID
+    status: Status = Status.INITIALIZED
+    name: str = 'cookgroup'
+    host_placement: HostPlacement = HostPlacement(HostPlacement.ALL)
+    straggler_handling: StragglerHandling = StragglerHandling()
+    jobs: List[UUID] = []
+
+    def __init__(self, *,
+                 uuid: Optional[UUID] = None,
+                 status: Status = Status.INITIALIZED,
+                 name: str = 'cookgroup',
+                 host_placement: HostPlacement = HostPlacement(
+                     HostPlacement.ALL),
+                 straggler_handling: StragglerHandling = StragglerHandling(),
+                 jobs: List[UUID] = []):
+        """Initialize a group.
+
+        :param uuid: The UUID of this group. If no UUID is specified, then one
+            will be generated with the `make_temporal_uuid()`.
+        :type uuid: Optional[UUID], optional
+        :param status: The status of this group, defaults to Status.INITIALIZED
+        :type status: Status, optional
+        :param name: A description of this group, defaults to 'cookgroup'
+        :type name: str, optional
+        :param host_placement: A host placement strategy for this group,
+            defaults to `HostPlacement(HostPlacement.ALL)`
+        :type host_placement: HostPlacement, optional
+        :param straggler_handling: A straggling job handling strategy,
+            defaults to the default straggler handling strategy.
+        :type straggler_handling: StragglerHandling, optional
+        :param jobs: UUIDs of the jobs this group includes, defaults to an
+            empty list.
+        :type jobs: List[UUID], optional
+        """
+        self.uuid = uuid if uuid is not None else util.make_temporal_uuid()
+        self.status = status
+        self.name = name
+        self.host_placement = host_placement
+        self.straggler_handling = straggler_handling
+        self.jobs = jobs
+
+    def to_dict(self) -> dict:
+        return {
+            'uuid': self.uuid,
+            'status': self.status,
+            'name': self.name,
+            'host_placement': self.host_placement,
+            'straggler_handling': self.straggler_handling,
+            'jobs': self.jobs
+        }
```

Comment:
My opinion is that we leave group support out of this PR as well to keep it small. What do you think?

Reply:
Makes sense, the `Group` class isn't instantiated anywhere at the moment so I can get rid of this